plutus-example/src/PlutusExample/ScriptContextChecker.hs

```diff
@@ -0,0 +1,433 @@
+import Cardano.Slotting.EpochInfo (EpochInfo, hoistEpochInfo)
+import Cardano.Slotting.Time (SystemStart)
+import Control.Monad.Trans.Except
+import Ledger qualified as Plutus
+import Ledger.Typed.Scripts qualified as Scripts
+import Ouroboros.Consensus.HardFork.Combinator.AcrossEras qualified as Consensus
+import Ouroboros.Consensus.HardFork.History qualified as Consensus
+import Ouroboros.Network.Protocol.LocalStateQuery.Type (AcquireFailure)
+import Plutus.V1.Ledger.DCert qualified as Plutus
+import PlutusTx qualified
+import PlutusTx.AssocMap qualified as AMap
+import PlutusTx.IsData.Class
+import PlutusTx.Prelude hiding (Semigroup (..), unless, (.))
+import PlutusTx.Prelude qualified as P
+
+-- Description
+-- MyCustomRedeemer mimics the ScriptContext. MyCustomRedeemer is built via reading
+-- the transaction containing the script and the script itself just compares MyCustomRedeemer
+-- to the ScriptContext to be sure they are equivalent.
+-- The overall aim is to make sure what is provided via ScriptContext (i.e. the transaction)
+-- is what it's supposed to be. We check this by creating MyCustomRedeemer based on
+-- the actual transaction which is created via the create-script-context executable.
+
+newtype MyCustomDatum = MyCustomDatum Integer
+
+data MyCustomRedeemer
+  = MyCustomRedeemer
+      { mCrOutputs       :: [Plutus.TxOut]
+      , mCrInputs        :: [Plutus.TxInInfo]
+      , mCrMint          :: Plutus.Value
+      , mCrValidRange    :: Plutus.POSIXTimeRange
+      , mCrFee           :: Plutus.Value
+      , mCrDatums        :: [(Plutus.DatumHash, Plutus.Datum)]
+      , mCrCerts         :: [Plutus.DCert]
+      , mCrSignatories   :: [Plutus.PubKeyHash]
+      , mCrScriptPurpose :: Maybe Plutus.ScriptPurpose
+      } deriving (Prelude.Eq, Show)
+
+PlutusTx.unstableMakeIsData ''MyCustomDatum
+PlutusTx.unstableMakeIsData ''MyCustomRedeemer
+
+data ScriptContextTest
+instance Scripts.ValidatorTypes ScriptContextTest where
+    type instance DatumType ScriptContextTest    = MyCustomDatum
+    type instance RedeemerType ScriptContextTest = MyCustomRedeemer
+
+{-# INLINABLE mkValidator #-}
+mkValidator :: MyCustomDatum-> MyCustomRedeemer -> Plutus.ScriptContext -> Bool
+mkValidator _datum (MyCustomRedeemer txouts txins minted txValidRange _fee datumsAndHashes certs signatories mPurpose) scriptContext =
+  -- Minted field is equivalent
+  Plutus.txInfoMint txInfo P.== minted P.&&
+  -- Validity range is equivalent
+  Plutus.txInfoValidRange txInfo P.== txValidRange P.&&
+  -- Datums and datum hashes are equivalent
+  Plutus.txInfoData txInfo P.== datumsAndHashes P.&&
+  -- Required tx signers are equivalent
+  Plutus.txInfoSignatories txInfo P.== signatories P.&&
+  -- Payment tx out is equivalent
+  AMap.member paymentOutputFromRedeemer scriptContextOutputsMap P.&&
+  -- Txins are equivalent
+  (AMap.member txinA scriptContextTxinsMap P.&& AMap.member txinB scriptContextTxinsMap) P.&&
+  -- Cert if equivalent
```

Comment:
Not clear to me what "cert if equivalent" means.  Does it mean "certify if equivalent"?  Or "certificate if equivalent?